File: crates/swc_ecma_transforms/scripts/fixtures.py
#!/usr/bin/env python3


from os import listdir
from os.path import isfile, join

files = [f for f in listdir('./fixtures') if isfile(join('./fixtures', f))]

fm = {}

for f in files:
    name = join('./fixtures', f)
    if '.DS_Store' in f:
        continue
    test_name, file_type = f.rsplit('_', 1)
    if file_type == 'input.mjs':
        file_type = 'input.js'
    if file_type == 'input.mjsz':
        file_type = 'input.js'
    if file_type == 'output.mjs':
        file_type = 'output.js'
    if file_type == 'output.mjsz':
        file_type = 'output.js'


    if file_type == 'exec.mjs':
        file_type = 'exec.js'
    if file_type == 'exec.mjsz':
        file_type = 'exec.js'

    if not test_name in fm:
        fm[test_name] = {}
    with open(name, "r") as f:
        fm[test_name][file_type] = f.read()


for name in fm:
    m = fm[name]
    name = name.replace('-', '_')

    print()
    print('// {}'.format(name))
    if 'exec.js' in m:
        if 'options.json' in m:
            print('test_exec!(syntax(), |_| tr(r#"{}"#), {}_exec, r#"\n{}\n"#);'.format(
                m['options.json'],name, m['exec.js']
            ))
        else:
            print('test_exec!(syntax(), |_| tr(Default::default()), {}_exec, r#"\n{}\n"#);'.format(
                name, m['exec.js']
            ))
    elif 'input.js' in m and 'output.js' in m:
        if 'options.json' in m:
            print('test!(syntax(),|_| tr(r#"{}"#), {}, r#"\n{}\n"#, r#"\n{}\n"#);'.format(
                m['options.json'], name, m['input.js'], m['output.js']
            ))
        else:
            print('test!(syntax(),|_| tr(Default::default()), {}, r#"\n{}\n"#, r#"\n{}\n"#);'.format(
                name, m['input.js'], m['output.js']
            ))
    elif 'stdout.txt' in m:
        pass
    else:
        print(m.keys())
        raise Exception(name, m)

File: crates/swc_ecma_transforms/scripts/fixtures.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const FIXTURE_DIR = path.join('.', 'fixtures');
const FILE_TYPE_ALIASES = Object.freeze({
  'exec.mjs': 'exec.js',
  'exec.mjsz': 'exec.js',
  'input.mjs': 'input.js',
  'input.mjsz': 'input.js',
  'output.mjs': 'output.js',
  'output.mjsz': 'output.js',
});

/**
 * Keep fixture emission deterministic so regenerated snippets are stable.
 *
 * The original helper tolerated Finder artifacts, so we keep ignoring them.
 */
const fixtureFiles = fs
  .readdirSync(FIXTURE_DIR, { withFileTypes: true })
  .filter((entry) => entry.isFile() && !entry.name.includes('.DS_Store'))
  .map((entry) => entry.name)
  .sort();

const fixturesByName = new Map();

for (const fileName of fixtureFiles) {
  const separatorIndex = fileName.lastIndexOf('_');
  if (separatorIndex === -1) {
    throw new Error(`invalid fixture file name: ${fileName}`);
  }

  const testName = fileName.slice(0, separatorIndex);
  const rawFileType = fileName.slice(separatorIndex + 1);
  const fileType = FILE_TYPE_ALIASES[rawFileType] ?? rawFileType;
  const filePath = path.join(FIXTURE_DIR, fileName);
  const contents = fs.readFileSync(filePath, 'utf8');

  if (!fixturesByName.has(testName)) {
    fixturesByName.set(testName, new Map());
  }

  fixturesByName.get(testName).set(fileType, contents);
}

for (const [rawName, files] of [...fixturesByName.entries()].sort(([left], [right]) =>
  left.localeCompare(right),
)) {
  const testName = rawName.replace(/-/g, '_');

  console.log();
  console.log(`// ${testName}`);

  if (files.has('exec.js')) {
    const execCode = files.get('exec.js');
    const options = files.get('options.json');

    if (options !== undefined) {
      console.log(
        `test_exec!(syntax(), |_| tr(r#"${options}"#), ${testName}_exec, r#"\\n${execCode}\\n"#);`,
      );
    } else {
      console.log(
        `test_exec!(syntax(), |_| tr(Default::default()), ${testName}_exec, r#"\\n${execCode}\\n"#);`,
      );
    }
  } else if (files.has('input.js') && files.has('output.js')) {
    const inputCode = files.get('input.js');
    const outputCode = files.get('output.js');
    const options = files.get('options.json');

    if (options !== undefined) {
      console.log(
        `test!(syntax(),|_| tr(r#"${options}"#), ${testName}, r#"\\n${inputCode}\\n"#, r#"\\n${outputCode}\\n"#);`,
      );
    } else {
      console.log(
        `test!(syntax(),|_| tr(Default::default()), ${testName}, r#"\\n${inputCode}\\n"#, r#"\\n${outputCode}\\n"#);`,
      );
    }
  } else if (files.has('stdout.txt')) {
    continue;
  } else {
    console.log([...files.keys()]);
    throw new Error(`unsupported fixture group: ${rawName}`);
  }
}
